src/realm/group_writer.cpp

```diff
@@ -301,25 +302,296 @@ GroupWriter::MapWindow* GroupWriter::get_window(ref_type start_ref, size_t size)
 
 #define REALM_ALLOC_DEBUG 0
 
+auto GroupWriter::map_reachable() -> std::vector<FreeSpaceEntry>
+{
+    class Collector : public Array::MemUsageHandler {
+
+    public:
+        void handle(ref_type ref, size_t, size_t used) override
+        {
+            blocks.emplace_back(FreeSpaceEntry{ref, used, current_version});
+        }
+        std::vector<FreeSpaceEntry> blocks;
+        uint64_t current_version;
+    };
+    // collect reachable blocks in all reachable versions
+    Collector collector;
+    auto kill_it = m_unreachable_versions.begin();
+    for (auto& entry : m_top_ref_map) {
+        collector.current_version = entry.first;
+        // skip any empty entries
+        if (entry.second.top_ref == 0)
+            continue;
+        // skip versions about to become unreachable
+        while (kill_it < m_unreachable_versions.end() && *kill_it < entry.first)
+            ++kill_it;
+        if (kill_it < m_unreachable_versions.end() && *kill_it == entry.first)
+            continue;
+        Array array(m_alloc);
+        array.init_from_ref(entry.second.top_ref);
+        array.report_memory_usage(collector);
+    }
+
+    std::sort(collector.blocks.begin(), collector.blocks.end(),
+              [](const FreeSpaceEntry& a, const FreeSpaceEntry& b) {
+                  if (a.ref == b.ref)
+                      return a.released_at_version < b.released_at_version;
+                  return a.ref < b.ref;
+              });
+    return collector.blocks;
+}
+
+void GroupWriter::backdate()
+{
+    struct FreeList {
+        Array positions;
+        Array lengths;
+        Array versions;
+        ref_type top_ref;
+        ref_type logical_file_size;
+        uint64_t version;
+        bool resolved = false;
+        bool present = false;
+        FreeList(Allocator& alloc, ref_type top, ref_type logical_file_size, uint64_t version)
+            : positions(alloc)
+            , lengths(alloc)
+            , versions(alloc)
+            , top_ref(top)
+            , logical_file_size(logical_file_size)
+            , version(version)
+        {
+        }
+    };
+
+
+    std::set<uint64_t> unreachables;
```

Comment:
Are the versions not sorted already. I what we want is a set, why not construct a set in the first place?

Reply:
Fixed